### ISSUE TITLE
Crafting recipes may be silently overwritten

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#4382]
Consider this Lua code:
```
for i=1,3 do
	minetest.register_craftitem("craft_bad:dummy"..i, {
		description = "Craft Testing Dummy Item "..i,
		inventory_image = "unknown_item",
		wield_image = "unknown_item",
	})
end
minetest.register_craft({
	recipe = {{"craft_bad:dummy3", "craft_bad:dummy3"}}, 
	output = "craft_bad:dummy1",
})
minetest.register_craft({
	recipe = {{"craft_bad:dummy3", "craft_bad:dummy3"}}, 
	output = "craft_bad:dummy2",
})
```

This registers 2 equal recipes with a different output. In other words, this should never happen in real Lua scripting.
If you execute this code, Minetest apparently **silently** overwrites the first crafting recipe, because if you try to to set the items into the crafting grid, you get `craft_bad:dummy2`. There is no warning, no error message and no log entry!

IMO this is a pretty serious issue as careless modders can accidentally destroy the crafting recipe of another mod and it is hard to even notice.

If this occours, I suggest at the very *least* to print an error-level debug message. If this kind of bug occours in actual modding of subgame creation, it might be very hard to even notice or to track this down.
The error message should be written as soon a recipe with the exact same ingredients is registered.

This issue occours as well with the types shapeless, fuel, cooking and toolrepair.

Tested in Minetest 0.4.14.

**Comment by rubenwardy** [comment]
What if the mod intends on overwriting a recipe?

**Comment by Wuzzy2** [comment]
Then the mod is evil, I guess. :D 
I think this is quite dangerous since it is potentionally error-prone.
Can you give me an example of a mod which actually does this intentnionally right now?

IMO chances that you accidentally overwrite a crafting recipe are much higher than you do so intentionally.

But I have an idea for your use case: Add a 3rd parameter (which is optional) for `minetest.register_craft`, called `overwrite`. If true,  Minetest is `told` that it is okay to overwrite existing recipes and will not issue an error message. In this case, I'd say it is fully in the modder's responsibility to make sure things don't go down the drain. ;-) It should not go without warnings in `lua_api.txt`.
If false (implied default value), Minetest will give the error as I suggested before.

Also, to keep things simple Minetest can just continue overwrite the recipe in *both* cases (like now), the main difference is that it screams “ERROR”.

**Comment by t4im** [comment]
Technic does that to deactivate or change crafts that mess with its game concept.

But i guess since `minetest.clear_craft(recipe)` was added, one could perhaps start **warning** about it, so such situations can be updated.

**Comment by paramat** [comment]
Seems expected behaviour.

**Comment by Wuzzy2** [comment]
How is this expected behaviour?
Can you explain how modders are supposed to work with this crafting system?

I suppose it is almost impossible in Minetest to register a crafting recipe without being sure there are no collisions.
Sure, you could use rubenwardy's script but this only checks against mods you have installed. If the collision happens with a mod you don't have, you're screwed although you think everything is fine. :-(
But strictly spoken, a modder would have to check for collisions against all other mods ever written, because any of those might be used together with your own mod. Now who exactly has ever downloaded every mod? ;-)
Therefore, whenever you add a crafting recipe, you can never be sure it wasn't already registered by another mod. IMO this is a sign of bad game design. Yes, I honestly think the whole crafting system is flawed by design, and this issue is one of the reasons. In a well-designed system, crafting recipe collisions would be logically impossible.
IMO it is not justified to blame the modder for recipe collisions as it is unfair to expect to check against all other mods ever written.

I fear the only way to fix this issue is by redesigning the crafting system. I have other reasons to dislike the crafting system, maybe I'll write something about it in the forums.


Tagging this as “feature request” is doing a great disservice to the underlying issue here.

**Comment by paramat** [comment]
Expected in the sense of a later registration overwriting an earlier one.
I agree to printing a warning about conflicting recipes, this is useful for other reasons, but not an error that halts the game though, as some mods may use this overwriting behaviour.
It's not a bug.

**Comment by Wuzzy2** [comment]
So you disagree that the crafting system is broken by design for allowing recipe collisions to happen so easily?

I suppose in 99% of cases modders do *not* want to overwrite recipes of others, because it breaks expected behaviour and causes overall weirdness and buggyness etc. It is rare that a modder *intentionally* wants to do this.
Also, haven't you read my concerns?
Even if you print some random warning, a recipe in some mod might still conflict with some other recipe from some other mod. So a warning is not helping much. Unless you have all mods installed (haha!) it is extremely hard to find such a recipe collision.
So I am still convinced the issue can not be fixed without a major overhaul of the crafting system itself.

**Comment by paramat** [comment]
I agree that overwriting is almost always not wanted, and that any mod that does this is a little sloppy and should be rewritten.

> a recipe in some mod might still conflict with some other recipe from some other mod you haven't installed with no warning given (of course). So a warning is not helping much. Unless you have all mods installed (haha!) it is extremely hard to find such a recipe collision.

Obviously, but it is impossibe to check recipes in mods not installed. A warning is useful because it applies to the mods currently installed and/or enabled.

A recipe conflict check could be intensive considering how there could be hundreds of recipes, so such a check should only be done at mod load, not during the game (some recipes may be registered in-game).

**Comment by t4im** [comment]
It would be better if the system simply allowed both recipes to coexist, so that crafting grids could offer the user selecting which output to craft.

There will be always recipe conflicts between mods.
Wouldn't you rather prefer more crafting flexibility to constant erring with certain mod combinations?

**Comment by paramat** [comment]
^ Nice idea, simpler and more lightweight.

Since conflicting recipes are only a problem if the corresponding mods are installed and enabled, any error should only occur for mods installed and enabled. For example you may have 2 conflicting recipes in 2 mods never used together.

**Comment by Wuzzy2** [comment]
Adding support for multiple possible outputs to choose from goes into the right direction. It not only will fix the issue, but also allows more crafting possibilities and modders don't have to crazy amounts of checking which can never be done anyway.
Also, it would make the crafting system as a whole a little bit less broken.

Now the question is how it should be implemented. By an “output grid”, maybe? Where each slot of the “output grid” works like the current singular output slot, but you only craft the item you picked from the particular slot.

What about replacement items? Should they go into output, too or should they stay in input like before? IMO I would say staying with input for now would be simpler and wouldn't cause too much headache, although I am not sure if it is the optimal solution.

**Comment by yer-mtd** [comment]
Bumping. moreblocks:stone_tile and xdecor:stone_tile are a good example of this. xdecor overwrites moreblocks recipe of the similar block - making moreblocks' variant of the block uncraftable while they have different application

**Comment by erlehmann** [comment]
I remember having encountered this in Mesecraft, where some items were not craftable because of this.

@Wuzzy2 could you maybe create a small mod that overwrites `minetest.register_craft()` to help debugging?
